### PR TITLE
Fix type of navigate method

### LINKED ISSUE #39158: gatsby-link 5.14.0 types broken
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Updating gatsby-link to 5.14.0 causes types to be broken. It appear's to be caused by #38590

See the kind of Typescript errors I'm receiving now:

```log
src/pages/signup/billing.tsx:102:14 - error TS2345: Argument of type 'string' is not assignable to parameter of type 'number'.

102     navigate('/signup/')
                 ~~~~~~~~~~
```

See the file `src/pages/redirect.tsx` in the reproduction link.

### Reproduction Link

https://codesandbox.io/p/devbox/serverless-hooks-xf3rhw?workspaceId=e7359dad-f2e3-4fc9-9dc7-a71e1e25e213

### Steps to Reproduce

1. Call `navigate` using a string

### Expected Result

Type checks pass

### Actual Result

Type checks fail with following error
```
Argument of type 'string' is not assignable to parameter of type 'number'.
```
### Environment

```shell
System:
    OS: macOS 15.0.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.11.0 - /var/folders/xh/ttknrcjs3tbfv99f83f454d40000gn/T/yarn--1731201934806-0.747145292822317/node
    Yarn: 1.22.21 - /var/folders/xh/ttknrcjs3tbfv99f83f454d40000gn/T/yarn--1731201934806-0.747145292822317/yarn
    npm: 10.9.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 130.0.6723.117
    Safari: 18.0.1
  npmPackages:
    gatsby: ^5.14.0 => 5.14.0
    gatsby-plugin-babel-plugin-typescript-to-proptypes: ^1.0.5 => 1.0.5
    gatsby-plugin-catch-links: ^5.14.0 => 5.14.0
    gatsby-plugin-google-gtag: ^5.14.0 => 5.14.0
    gatsby-plugin-image: ^3.14.0 => 3.14.0
    gatsby-plugin-manifest: ^5.14.0 => 5.14.0
    gatsby-plugin-offline: ^6.14.0 => 6.14.0
    gatsby-plugin-perf-budgets: ^0.0.18 => 0.0.18
    gatsby-plugin-robots-txt: ^1.8.0 => 1.8.0
    gatsby-plugin-sass: ^6.14.0 => 6.14.0
    gatsby-plugin-sharp: ^5.14.0 => 5.14.0
    gatsby-plugin-sitemap: ^6.14.0 => 6.14.0
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.32 => 1.1.32
    gatsby-remark-autolink-headers: ^6.14.0 => 6.14.0
    gatsby-remark-external-links: ^0.0.4 => 0.0.4
    gatsby-remark-images: ^7.14.0 => 7.14.0
    gatsby-remark-smartypants: ^6.14.0 => 6.14.0
    gatsby-remark-table-of-contents: ^2.0.0 => 2.0.0
    gatsby-remark-vscode: ^3.3.1 => 3.3.1
    gatsby-source-filesystem: ^5.14.0 => 5.14.0
    gatsby-transformer-inline-svg: ^1.2.0 => 1.2.0
    gatsby-transformer-remark: ^6.14.0 => 6.14.0
    gatsby-transformer-sharp: ^5.14.0 => 5.14.0
    gatsby-transformer-yaml: ^5.14.0 => 5.14.0
```

### Config Flags

_No response_

### PR DESCRIPTION
**Pull request by pajosieg**
## Description

The usage of the original navigation function type created problems when using spreaded parameters. Since the navigate implementation in gatsby does not have the same call signature as @reach/router anyways (the promise returned by @reach/router is ignored), it is easier to define our own interface than to overwrite the types from @reach/router.

## Related Issues

Fixes #39158

### CODE REVIEW

**Comment by pajosieg** [comment]
@pieh does it matter that most tests fail? These are not related to the fix

**Comment by abemedia** [comment]
Would love to see this merged as this is currently blocking me from updating Gatsby.

File: packages/gatsby-link/index.d.ts [hunk 1]
```diff
@@ -33,7 +33,10 @@ export class Link<TState> extends React.Component<
  * Sometimes you need to navigate to pages programmatically, such as during form submissions. In these
  * cases, `Link` won’t work.
  */
-export const navigate: (...args: Parameters<NavigateFn>) => void;
+export const navigate: {
+  (to: string, options?: NavigateOptions<{}>): void
+  (to: number): void
+}
```
Comment (serhalp) [inline packages/gatsby-link/index.d.ts:39]: ref: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3fe735cfa0d0d25922b71d02ac748a96ea3d67a8/types/reach__router/index.d.ts#L103-L106

👍🏼 Looks right

**Review by serhalp — APPROVED** [review]
LGTM. Thank you for contributing a fix!

**Comment by pieh** [comment]
Published:
```
 - gatsby-link@5.14.1
 - gatsby@5.14.1
```

**Pull request merged**